dbsnap_verify/__init__.py

```diff
@@ -0,0 +1,187 @@
+from .rds_funcs import (
+    get_latest_snapshot,
+    get_database_description,
+    restore_from_latest_snapshot,
+    modify_db_instance_for_verify,
+    destroy_database,
+    destroy_database_subnet_group,
+    rds_event_messages,
+)
+
+from .state_doc import (
+    current_state,
+    transition_state,
+    get_or_create_state_doc,
+    clean_state_doc,
+)
+
+import boto3
+
+# retry 3 times on errors.
+from botocore.config import Config
+BOTO3_CONFIG = Config(retries={"max_attempts":3})
+
+import logging
+logger = logging.getLogger(__name__)
+
+try:
+    basestring
+except NameError:
+    basestring = str
+
+
+def wait(state_doc, rds_session):
+    """wait: currently waiting for the next snapshot to appear."""
+    verified_snapshot_id = state_doc.get("snapshot_verified", "any snapshot")
+    logger.info(
+        "Looking for a snapshot of %s (newer then %s)",
+        state_doc["database"],
+        verified_snapshot_id,
+    )
+    snapshot_desc = get_latest_snapshot(rds_session, state_doc["database"])
+    if snapshot_desc["DBSnapshotIdentifier"] != verified_snapshot_id:
+        # if the latest snapshot is older then the minimum, restore it.
+        state_doc["snapshot_verifying"] = snapshot_desc["DBSnapshotIdentifier"]
+        transition_state(state_doc, "restore")
+        restore(state_doc, rds_session)
+    else:
+        # continue waiting for a new snapshot to restore.
+        transition_state(state_doc, "wait")
+        logger.info(
+            "Did not find a snapshot of %s (newer then %s)",
+            state_doc["database"],
+            verified_snapshot_id,
+        )
+        logger.info("Going to sleep.")
+
+
+def restore(state_doc, rds_session):
+    """restore: currently restoring a copy of the latest
+    snapshot into a temporary RDS db instance."""
+    tmp_db_description = get_database_description(
+        rds_session, state_doc["tmp_database"]
+    )
+    if tmp_db_description is None:
+        logger.info(
+            "Restoring snapshot of %s to %s",
+            state_doc["database"],
+            state_doc["tmp_database"]
+        )
+        subnet_ids = state_doc["database_subnet_ids"]
+        if isinstance(subnet_ids, basestring):
+            subnet_ids = subnet_ids.split(",")
+        restore_from_latest_snapshot(
+            rds_session, state_doc["database"], subnet_ids
+        )
+    elif tmp_db_description["DBInstanceStatus"] == "available":
+        transition_state(state_doc, "modify")
+        modify(state_doc, rds_session)
+
+
+def modify(state_doc, rds_session):
+    """modify: currently modifying the temporary RDS db instance
+    settings to allow the dbsnap-verify tool to access it."""
+    logger.info(
+        "Modifying %s master password and security groups",
+        state_doc["tmp_database"]
+    )
+    security_group_ids = state_doc["database_security_group_ids"]
+    if isinstance(security_group_ids, basestring):
```

Comment:
Seems like this happens often enough that it should probably be a helper function.

Reply:
Or maybe the state_doc should handle it automatically?

Reply:
Fixed see PR #3